### PR TITLE
Indexer: cross-validations script for checkpoint, txn, object and event.

### PR DESCRIPTION
**Pull request by gegaowp**
## Description 

Hit FN and indexer RPC server separately for data consistency check.

## Test Plan 

local cargo run checking all checkpoints, txns, objects and events. Note that locally I let indexer forward all requests, so right now it's not checking the indexer-native impl yet.
```
cargo run --package sui-indexer --bin indexer_data_validation -- --fn-rpc-client-url "http://0.0.0.0:9000" --indexer-rpc-client-url 'http://127.0.0.1:3030' --start-checkpoint 0 --end-checkpoint 5 --check-transactions --check-events --check-objects
    Finished dev [unoptimized + debuginfo] target(s) in 0.57s
     Running `target/debug/indexer_data_validation --fn-rpc-client-url 'http://0.0.0.0:9000' --indexer-rpc-client-url 'http://127.0.0.1:3030' --start-checkpoint 0 --end-checkpoint 5 --check-transactions --check-events --check-objects`
2023-03-21T02:50:44.633633Z  INFO indexer_data_validation: Running correctness check for indexer...
2023-03-21T02:50:44.635310Z  INFO sui_indexer: Getting new RPC client...
2023-03-21T02:50:44.656130Z  INFO sui_indexer: Getting new RPC client...
2023-03-21T02:50:44.674736Z  INFO indexer_data_validation: Checking checkpoint 0
2023-03-21T02:50:45.024190Z  INFO indexer_data_validation: Checking checkpoint 1
2023-03-21T02:50:45.036347Z  INFO indexer_data_validation: Checking checkpoint 2
2023-03-21T02:50:45.049233Z  INFO indexer_data_validation: Checking checkpoint 3
2023-03-21T02:50:45.061343Z  INFO indexer_data_validation: Checking checkpoint 4
2023-03-21T02:50:45.072935Z  INFO indexer_data_validation: Checking checkpoint 5
```

### CODE REVIEW

File: crates/sui-indexer/src/bin/indexer_data_validation.rs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        .get_latest_checkpoint_sequence_number()
+        .await?;
+
+    let checkpoint = config.start_checkpoint;
+    if checkpoint <= fn_latest_checkpoint && checkpoint <= indexer_latest_checkpoint {
+        // TODO(gegaowp): think of schedule and coordination with checkpoint and epoch,
+        // now try to check past 10 checkpoints.
+        for checkpoint in max(checkpoint - 10, 0)..checkpoint {
+            check_checkpoint(&config, &fn_rpc_client, &indexer_rpc_client, checkpoint).await?;
+        }
```
Comment (666lcz) [inline crates/sui-indexer/src/bin/indexer_data_validation.rs:40]: nit: 10 checkpoint is like less than 1 minute. maybe add a `end_checkpoint`?

File: crates/sui-indexer/src/bin/indexer_data_validation.rs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        .read_api()
+        .get_checkpoint(CheckpointId::SequenceNumber(checkpoint))
+        .await?;
+
+    if fn_checkpoint != indexer_checkpoint {
+        error!("Checkpoint mismatch found in {}", checkpoint);
+        warn!(
+            "Checkpoint mismatch with checkpoint seq {:?}:\nFN:\n{:?}\nIndexer:\n{:?} ",
+            checkpoint, fn_checkpoint, indexer_checkpoint
+        );
```
Comment (666lcz) [inline crates/sui-indexer/src/bin/indexer_data_validation.rs:68]: should we just pick one?

**Review by 666lcz — APPROVED** [review]
accept to unblock

File: crates/sui-indexer/src/bin/indexer_data_validation.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+use sui_json_rpc_types::{
+    CheckpointId, ObjectChange, SuiObjectDataOptions, SuiTransactionResponseOptions,
+};
+use sui_sdk::SuiClient;
+use sui_types::base_types::{ObjectID, SequenceNumber};
+use sui_types::digests::TransactionDigest;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    info!("Running correctness check for indexer...");
```
Comment (666lcz) [inline crates/sui-indexer/src/bin/indexer_data_validation.rs:20]: Need to add this for the log to show up in terminal

```suggestion
async fn main() -> Result<()> {
    // Initialize logging
    let (_guard, filter_handle) = telemetry_subscribers::TelemetryConfig::new()
      .with_env()
      .with_prom_registry(&prometheus_registry)
      .init();
    info!("Running correctness check for indexer...");
```

File: crates/sui-indexer/src/bin/indexer_data_validation.rs (hunk 2 above)
Comment (gegaowp) [reply crates/sui-indexer/src/bin/indexer_data_validation.rs:69]: I wanted to use them differently:
- error can be used as "title" or counts only
- warn is for deeper investigations

**Pull request merged**